### PR TITLE
feat: ButtonWrapper prop in Button to customise rendering

### PR DESCRIPTION
**Pull request by jsomsanith**
**What is the problem this PR is trying to solve?**
This PR resolve PR #36.
The Link component has a `LinkWrapper` feature, that takes a component to render, adding a Link style on it.
We want the same feature for button.

**What is the chosen solution to this problem?**
`Button` component now has a `ButtonWrapper` props, allowing to write that :
```
<Button ButtonWrapper={CustomButton} appearance="primary">
        Primary
</Button>

<Button ButtonWrapper={CustomAnchor} appearance="primary" href="http://storybook.js.org">
        Primary
</Button>
```

This will render `CustomButton`/`CustomAnchor` applying the `Button` style

### CODE REVIEW

**Comment by domyen** [comment]
Awesome! Can you update the Button story so folks can find out about the usage?

**Comment by jsomsanith** [comment]
Oh, I added 2 stories, guess I haven’t pushed them

**Comment by jsomsanith** [comment]
That’s it :D I just pushed

**Comment by jsomsanith** [comment]
@kylesuss fixed, and I added an optimisation: we memoize the styled button wrapper, instead of creating a new component at each render. Everytime it’s a new component, React considers it as a component change, the old will always be unmounted, and a new one mounted instead.

File: src/components/Button.js [hunk 1]
```diff
@@ -305,27 +305,46 @@ const ButtonWrapper = styled.button`
...
 `;
 
-const ButtonLink = ButtonWrapper.withComponent('a');
+const ButtonLink = StyledButton.withComponent('a');
 
-export function Button({ isDisabled, isLoading, loadingText, isLink, children, ...props }) {
+const applyStyle = ButtonWrapper => {
+  return (
+    ButtonWrapper &&
+    StyledButton.withComponent(({ containsIcon, isLoading, isUnclickable, ...rest }) => (
```
Comment (kylesuss) [inline src/components/Button.js]: Ah cool! Didn't realize you could do this same destructuring technique using `withComponent`.

File: src/components/Button.js [hunk 2]
```diff
@@ -305,27 +305,46 @@ const ButtonWrapper = styled.button`
...
     </Fragment>
   );
 
-  if (isLink) {
-    return (
-      <ButtonLink isLoading={isLoading} disabled={isDisabled} {...props}>
-        {buttonInner}
-      </ButtonLink>
-    );
+  const StyledButtonWrapper = React.useMemo(() => applyStyle(ButtonWrapper), [ButtonWrapper]);
```
Comment (kylesuss) [inline src/components/Button.js]: Nice idea w/ the memoization 👌

**Comment by kylesuss** [comment]
:rocket: PR was released in v0.0.35 :rocket:

**Pull request merged**